File: client/ios/App/App/CallManager.swift
import Foundation
import WebRTC
import AVFoundation

enum CallKind: String {
    case audio
    case video
}

struct CallPeer: Equatable {
    let id: String
    let displayName: String
    let avatarColor: String?
    let avatarUrl: String?
}

enum CallState: Equatable {
    case idle
    case outgoing(peer: CallPeer, kind: CallKind)
    case incoming(callId: String, peer: CallPeer, conversationId: String, kind: CallKind)
    case connecting(peer: CallPeer, kind: CallKind)
    case connected(peer: CallPeer, kind: CallKind, startedAt: Date)

    static func == (lhs: CallState, rhs: CallState) -> Bool {
        switch (lhs, rhs) {
        case (.idle, .idle): return true
        case (.outgoing(let a, let ka), .outgoing(let b, let kb)): return a == b && ka == kb
        case (.incoming(let ca, let pa, let coa, let ka), .incoming(let cb, let pb, let cob, let kb)):
            return ca == cb && pa == pb && coa == cob && ka == kb
        case (.connecting(let a, let ka), .connecting(let b, let kb)): return a == b && ka == kb
        case (.connected(let a, let ka, _), .connected(let b, let kb, _)): return a == b && ka == kb
        default: return false
        }
    }
}

final class CallManager: NSObject, ObservableObject {
    static let shared = CallManager()

    @Published private(set) var state: CallState = .idle
    @Published private(set) var localVideoTrack: RTCVideoTrack?
    @Published private(set) var remoteVideoTrack: RTCVideoTrack?
    @Published var isMuted = false
    @Published var isSpeakerOn = true
    @Published var isVideoEnabled = true
    @Published var lastError: String?

    private var webRTCClient: WebRTCClient?
    private var currentCallId: String?
    private var targetUserId: String?
    private var pendingRemoteCandidates: [RTCIceCandidate] = []
    private var remoteDescriptionSet = false
    private var pendingInviteSDP: SDPPayload?

    private override init() {
        super.init()
    }

    func registerSignaling() {
        AppSocketManager.shared.onCallInvite = { [weak self] event in
            self?.handleInvite(event)
        }
        AppSocketManager.shared.onCallAnswer = { [weak self] event in
            self?.handleAnswer(event)
        }
        AppSocketManager.shared.onCallIceCandidate = { [weak self] event in
            self?.handleRemoteCandidate(event)
        }
        AppSocketManager.shared.onCallReject = { [weak self] event in
            self?.handleReject(event)
        }
        AppSocketManager.shared.onCallEnd = { [weak self] event in
            self?.handleRemoteEnd(event)
        }
    }

    private func showError(_ message: String) {
        lastError = message
        DispatchQueue.main.asyncAfter(deadline: .now() + 6) { [weak self] in
            if self?.lastError == message {
                self?.lastError = nil
            }
        }
    }

    // MARK: Permissions

    private func checkCallPermissions(needsVideo: Bool, completion: @escaping (Bool) -> Void) {
        AVAudioSession.sharedInstance().requestRecordPermission { [weak self] audioGranted in
            DispatchQueue.main.async {
                guard let self else { return }
                guard audioGranted else {
                    self.showError("Нет доступа к микрофону. Разрешите его в Настройки → Wave → Микрофон.")
                    completion(false)
                    return
                }
                guard needsVideo else {
                    completion(true)
                    return
                }
                switch AVCaptureDevice.authorizationStatus(for: .video) {
                case .authorized:
                    completion(true)
                case .notDetermined:
                    AVCaptureDevice.requestAccess(for: .video) { videoGranted in
                        DispatchQueue.main.async {
                            if !videoGranted {
                                self.showError("Нет доступа к камере. Разрешите его в Настройки → Wave → Камера.")
                            }
                            completion(videoGranted)
                        }
                    }
                default:
                    self.lastError = "Нет доступа к камере. Разрешите его в Настройки → Wave → Камера."
                    completion(false)
                }
            }
        }
    }

    // MARK: Outgoing

    func startCall(conversation: Conversation, kind: CallKind) {
        guard case .idle = state, let otherUser = conversation.otherUser else { return }
        checkCallPermissions(needsVideo: kind == .video) { [weak self] granted in
            guard let self, granted else { return }
            self.beginOutgoingCall(conversation: conversation, otherUser: otherUser, kind: kind)
        }
    }

    private func beginOutgoingCall(conversation: Conversation, otherUser: Member, kind: CallKind) {
        let callId = UUID().uuidString
        currentCallId = callId
        targetUserId = otherUser.id
        isMuted = false
        isVideoEnabled = true
        isSpeakerOn = kind == .video

        let peer = CallPeer(id: otherUser.id, displayName: otherUser.displayName, avatarColor: otherUser.avatarColor, avatarUrl: otherUser.avatarUrl)
        state = .outgoing(peer: peer, kind: kind)

        let client = WebRTCClient(isVideoCall: kind == .video)
        client.delegate = self
        webRTCClient = client
        localVideoTrack = client.localVideoTrack
        if kind == .video {
            client.speakerOn()
            client.startCapturingVideo()
        }

        client.createOffer { [weak self] sdp in
            guard let self, let targetUserId = self.targetUserId, let callId = self.currentCallId else { return }
            AppSocketManager.shared.sendCallInvite(
                conversationId: conversation.id,
                targetUserId: targetUserId,
                callId: callId,
                kind: kind.rawValue,
                sdp: sdp
            )
        }
    }

    // MARK: Incoming

    func acceptCall() {
        guard case .incoming(_, _, _, let kind) = state else { return }
        checkCallPermissions(needsVideo: kind == .video) { [weak self] granted in
            guard let self else { return }
            if granted {
                self.beginAcceptCall()
            } else {
                self.rejectCall()
            }
        }
    }

    private func beginAcceptCall() {
        guard case .incoming(let callId, let peer, _, let kind) = state, let sdp = pendingInviteSDP else { return }
        targetUserId = peer.id
        currentCallId = callId
        isMuted = false
        isVideoEnabled = true
        isSpeakerOn = kind == .video

        let client = WebRTCClient(isVideoCall: kind == .video)
        client.delegate = self
        webRTCClient = client
        localVideoTrack = client.localVideoTrack
        if kind == .video {
            client.speakerOn()
            client.startCapturingVideo()
        }

        state = .connecting(peer: peer, kind: kind)

        let remoteSdp = RTCSessionDescription(type: WebRTCClient.sdpType(from: sdp.type), sdp: sdp.sdp)
        client.setRemoteDescription(remoteSdp) { [weak self] _ in
            guard let self else { return }
            self.remoteDescriptionSet = true
            self.flushPendingCandidates()
            client.createAnswer { sdp in
                AppSocketManager.shared.sendCallAnswer(targetUserId: peer.id, callId: callId, sdp: sdp)
            }
        }
    }

    func rejectCall() {
        guard case .incoming(let callId, let peer, _, _) = state else { return }
        AppSocketManager.shared.sendCallReject(targetUserId: peer.id, callId: callId)
        resetState()
    }

    // MARK: Ending

    func endCall() {
        if let targetUserId, let callId = currentCallId {
            AppSocketManager.shared.sendCallEnd(targetUserId: targetUserId, callId: callId)
        }
        resetState()
    }

    // MARK: Controls

    func toggleMute() {
        isMuted.toggle()
        webRTCClient?.setAudioEnabled(!isMuted)
    }

    func toggleSpeaker() {
        isSpeakerOn.toggle()
        if isSpeakerOn {
            webRTCClient?.speakerOn()
        } else {
            webRTCClient?.speakerOff()
        }
    }

    func toggleVideo() {
        isVideoEnabled.toggle()
        webRTCClient?.setVideoEnabled(isVideoEnabled)
    }

    // MARK: Signaling handlers

    private func handleInvite(_ event: CallInviteEvent) {
        guard case .idle = state else {
            AppSocketManager.shared.sendCallReject(targetUserId: event.fromUserId, callId: event.callId)
            return
        }
        pendingInviteSDP = event.sdp
        let peer = CallPeer(
            id: event.fromUserId,
            displayName: (event.fromDisplayName?.isEmpty == false) ? event.fromDisplayName! : "Собеседник",
            avatarColor: event.fromAvatarColor,
            avatarUrl: event.fromAvatarUrl
        )
        state = .incoming(
            callId: event.callId,
            peer: peer,
            conversationId: event.conversationId,
            kind: CallKind(rawValue: event.kind) ?? .audio
        )
    }

    private func handleAnswer(_ event: CallAnswerEvent) {
        guard event.callId == currentCallId, let client = webRTCClient else { return }
        let sdp = RTCSessionDescription(type: WebRTCClient.sdpType(from: event.sdp.type), sdp: event.sdp.sdp)
        client.setRemoteDescription(sdp) { [weak self] _ in
            guard let self else { return }
            self.remoteDescriptionSet = true
            self.flushPendingCandidates()
            DispatchQueue.main.async {
                if case .outgoing(let peer, let kind) = self.state {
                    self.state = .connecting(peer: peer, kind: kind)
                }
            }
        }
    }

    private func handleRemoteCandidate(_ event: CallIceCandidateEvent) {
        guard event.callId == currentCallId else { return }
        let candidate = RTCIceCandidate(
            sdp: event.candidate.candidate,
            sdpMLineIndex: event.candidate.sdpMLineIndex,
            sdpMid: event.candidate.sdpMid
        )
        if remoteDescriptionSet, let client = webRTCClient {
            client.addIceCandidate(candidate)
        } else {
            pendingRemoteCandidates.append(candidate)
        }
    }

    private func handleReject(_ event: CallRejectEvent) {
        guard event.callId == currentCallId else { return }
        resetState()
    }

    private func handleRemoteEnd(_ event: CallEndEvent) {
        guard event.callId == currentCallId else { return }
        resetState()
    }

    private func flushPendingCandidates() {
        guard let client = webRTCClient else { return }
        for candidate in pendingRemoteCandidates {
            client.addIceCandidate(candidate)
        }
        pendingRemoteCandidates.removeAll()
    }

    private func resetState() {
        webRTCClient?.close()
        webRTCClient = nil
        currentCallId = nil
        targetUserId = nil
        pendingRemoteCandidates.removeAll()
        remoteDescriptionSet = false
        pendingInviteSDP = nil
        localVideoTrack = nil
        remoteVideoTrack = nil
        isMuted = false
        state = .idle
    }
}

extension CallManager: WebRTCClientDelegate {
    func webRTCClient(_ client: WebRTCClient, didGenerateCandidate candidate: RTCIceCandidate) {
        guard let targetUserId, let callId = currentCallId else { return }
        AppSocketManager.shared.sendCallIceCandidate(targetUserId: targetUserId, callId: callId, candidate: candidate)
    }

    func webRTCClient(_ client: WebRTCClient, didReceiveRemoteVideoTrack track: RTCVideoTrack) {
        DispatchQueue.main.async { [weak self] in
            self?.remoteVideoTrack = track
        }
    }

    func webRTCClient(_ client: WebRTCClient, didChangeConnectionState connectionState: RTCIceConnectionState) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            switch connectionState {
            case .connected, .completed:
                if case .connecting(let peer, let kind) = self.state {
                    self.state = .connected(peer: peer, kind: kind, startedAt: Date())
                } else if case .outgoing(let peer, let kind) = self.state {
                    self.state = .connected(peer: peer, kind: kind, startedAt: Date())
                }
            case .disconnected, .failed, .closed:
                self.resetState()
            default:
                break
            }
        }
    }
}

File: android-native/app/src/main/java/com/wave/app/call/CallManager.kt
package com.wave.app.call

import android.content.Context
import android.media.AudioManager
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wave.app.data.SessionStore
import com.wave.app.model.Conversation
import com.wave.app.network.CallAnswerEvent
import com.wave.app.network.CallEndEvent
import com.wave.app.network.CallIceCandidateEvent
import com.wave.app.network.CallInviteEvent
import com.wave.app.network.SdpPayload
import com.wave.app.network.SocketManager
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.SessionDescription
import org.webrtc.VideoTrack
import java.util.UUID

enum class CallKind { AUDIO, VIDEO }

data class CallPeer(val id: String, val displayName: String, val avatarColor: String?, val avatarUrl: String?)

sealed class CallState {
    object Idle : CallState()
    data class Outgoing(val peer: CallPeer, val kind: CallKind) : CallState()
    data class Incoming(val callId: String, val peer: CallPeer, val conversationId: String, val kind: CallKind) : CallState()
    data class Connecting(val peer: CallPeer, val kind: CallKind) : CallState()
    data class Connected(val peer: CallPeer, val kind: CallKind, val startedAt: Long) : CallState()
}

/**
 * App-wide call orchestrator, mirroring the iOS CallManager singleton:
 * one active call at a time, client owns the whole call lifecycle, the
 * server only relays SDP/ICE by target user id.
 */
object CallManager : WebRTCClientListener {
    var state by mutableStateOf<CallState>(CallState.Idle)
        private set
    var localVideoTrack by mutableStateOf<VideoTrack?>(null)
        private set
    var remoteVideoTrack by mutableStateOf<VideoTrack?>(null)
        private set
    var isMuted by mutableStateOf(false)
    var isSpeakerOn by mutableStateOf(true)
    var isVideoEnabled by mutableStateOf(true)

    private lateinit var appContext: Context
    private var webRTCClient: WebRTCClient? = null
    private var currentCallId: String? = null
    private var targetUserId: String? = null
    private val pendingCandidates = mutableListOf<IceCandidate>()
    private var remoteDescriptionSet = false
    private var pendingInviteSdp: SdpPayload? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private fun runOnMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) block() else mainHandler.post(block)
    }

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    fun registerSignaling() {
        SocketManager.onCallInvite = { handleInvite(it) }
        SocketManager.onCallAnswer = { handleAnswer(it) }
        SocketManager.onCallIceCandidate = { handleRemoteCandidate(it) }
        SocketManager.onCallReject = { handleReject(it) }
        SocketManager.onCallEnd = { handleRemoteEnd(it) }
    }

    fun startCall(conversation: Conversation, kind: CallKind) {
        val other = conversation.otherUser ?: return
        if (state != CallState.Idle) return
        val callId = UUID.randomUUID().toString()
        currentCallId = callId
        targetUserId = other.id
        isMuted = false
        isVideoEnabled = true
        isSpeakerOn = kind == CallKind.VIDEO
        val peer = CallPeer(other.id, other.displayName, other.avatarColor, other.avatarUrl)
        state = CallState.Outgoing(peer, kind)

        val client = setupClient(kind)
        if (kind == CallKind.VIDEO) client.startCapturingVideo()
        applyAudioRouting()

        client.createOffer { sdp ->
            val user = SessionStore(appContext).user
            SocketManager.sendCallInvite(
                conversationId = conversation.id,
                targetUserId = other.id,
                callId = callId,
                kind = kind.name.lowercase(),
                sdp = sdp,
                fromDisplayName = user?.displayName,
                fromAvatarColor = user?.avatarColor,
                fromAvatarUrl = user?.avatarUrl
            )
        }
    }

    fun acceptCall() {
        val current = state as? CallState.Incoming ?: return
        val sdp = pendingInviteSdp ?: return
        targetUserId = current.peer.id
        isMuted = false
        isVideoEnabled = true
        isSpeakerOn = current.kind == CallKind.VIDEO
        state = CallState.Connecting(current.peer, current.kind)

        val client = setupClient(current.kind)
        if (current.kind == CallKind.VIDEO) client.startCapturingVideo()
        applyAudioRouting()

        val remoteSdp = SessionDescription(WebRTCClient.sdpType(sdp.type), sdp.sdp)
        client.setRemoteDescription(remoteSdp) {
            runOnMain {
                remoteDescriptionSet = true
                flushPendingCandidates()
                client.createAnswer { answer ->
                    SocketManager.sendCallAnswer(current.peer.id, current.callId, answer)
                }
            }
        }
    }

    fun rejectCall() {
        val current = state as? CallState.Incoming ?: return
        SocketManager.sendCallReject(current.peer.id, current.callId)
        resetState()
    }

    fun endCall() {
        val callId = currentCallId
        val target = targetUserId
        if (target != null && callId != null) {
            SocketManager.sendCallEnd(target, callId)
        }
        resetState()
    }

    fun toggleMute() {
        isMuted = !isMuted
        webRTCClient?.setAudioEnabled(!isMuted)
    }

    fun toggleSpeaker() {
        isSpeakerOn = !isSpeakerOn
        applyAudioRouting()
    }

    fun toggleVideo() {
        isVideoEnabled = !isVideoEnabled
        webRTCClient?.setVideoEnabled(isVideoEnabled)
    }

    private fun applyAudioRouting() {
        val audioManager = appContext.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return
        audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
        audioManager.isSpeakerphoneOn = isSpeakerOn
    }

    private fun setupClient(kind: CallKind): WebRTCClient {
        val client = WebRTCClient(appContext, kind == CallKind.VIDEO, this)
        webRTCClient = client
        localVideoTrack = client.localVideoTrack
        return client
    }

    private fun handleInvite(event: CallInviteEvent) = runOnMain {
        if (state != CallState.Idle) {
            SocketManager.sendCallReject(event.fromUserId, event.callId)
            return@runOnMain
        }
        pendingInviteSdp = event.sdp
        val peer = CallPeer(
            id = event.fromUserId,
            displayName = event.fromDisplayName?.takeIf { it.isNotEmpty() } ?: "Собеседник",
            avatarColor = event.fromAvatarColor,
            avatarUrl = event.fromAvatarUrl
        )
        state = CallState.Incoming(
            callId = event.callId,
            peer = peer,
            conversationId = event.conversationId,
            kind = if (event.kind == "video") CallKind.VIDEO else CallKind.AUDIO
        )
    }

    private fun handleAnswer(event: CallAnswerEvent) {
        val client = webRTCClient ?: return
        if (event.callId != currentCallId) return
        val sdp = SessionDescription(WebRTCClient.sdpType(event.sdp.type), event.sdp.sdp)
        client.setRemoteDescription(sdp) {
            runOnMain {
                remoteDescriptionSet = true
                flushPendingCandidates()
                (state as? CallState.Outgoing)?.let { state = CallState.Connecting(it.peer, it.kind) }
            }
        }
    }

    private fun handleRemoteCandidate(event: CallIceCandidateEvent) {
        if (event.callId != currentCallId) return
        val candidate = IceCandidate(event.candidate.sdpMid, event.candidate.sdpMLineIndex, event.candidate.candidate)
        if (remoteDescriptionSet) {
            webRTCClient?.addIceCandidate(candidate)
        } else {
            pendingCandidates.add(candidate)
        }
    }

    private fun handleReject(event: CallEndEvent) = runOnMain {
        if (event.callId == currentCallId) resetState()
    }

    private fun handleRemoteEnd(event: CallEndEvent) = runOnMain {
        if (event.callId == currentCallId) resetState()
    }

    private fun flushPendingCandidates() {
        val client = webRTCClient ?: return
        pendingCandidates.forEach { client.addIceCandidate(it) }
        pendingCandidates.clear()
    }

    private fun resetState() {
        webRTCClient?.close()
        webRTCClient = null
        currentCallId = null
        targetUserId = null
        pendingCandidates.clear()
        remoteDescriptionSet = false
        pendingInviteSdp = null
        localVideoTrack = null
        remoteVideoTrack = null
        isMuted = false
        state = CallState.Idle
        if (::appContext.isInitialized) {
            val audioManager = appContext.getSystemService(Context.AUDIO_SERVICE) as? AudioManager
            audioManager?.mode = AudioManager.MODE_NORMAL
            audioManager?.isSpeakerphoneOn = false
        }
    }

    // MARK: WebRTCClientListener

    override fun onIceCandidate(candidate: IceCandidate) {
        val target = targetUserId ?: return
        val callId = currentCallId ?: return
        SocketManager.sendCallIceCandidate(target, callId, candidate)
    }

    override fun onConnectionStateChange(state: PeerConnection.IceConnectionState) = runOnMain {
        when (state) {
            PeerConnection.IceConnectionState.CONNECTED, PeerConnection.IceConnectionState.COMPLETED -> {
                val current = this.state
                if (current is CallState.Connecting) {
                    this.state = CallState.Connected(current.peer, current.kind, System.currentTimeMillis())
                } else if (current is CallState.Outgoing) {
                    this.state = CallState.Connected(current.peer, current.kind, System.currentTimeMillis())
                }
            }
            PeerConnection.IceConnectionState.DISCONNECTED,
            PeerConnection.IceConnectionState.FAILED,
            PeerConnection.IceConnectionState.CLOSED -> resetState()
            else -> {}
        }
    }

    override fun onRemoteVideoTrack(track: VideoTrack) = runOnMain {
        remoteVideoTrack = track
    }
}
